File: basic_dsa_one_hundred_sixty_one.java
/**
 * Basic DSA Problems - Binary Tree Maximum Depth
 * Height of binary tree (longest path from root to leaf)
 */
public class basic_dsa_one_hundred_sixty_one {

    // Problem: Given root of binary tree, return its maximum depth (number of nodes on longest path).

    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    /**
     * Depth = 1 + max(depth(left), depth(right)). Base: null -> 0.
     * Time O(n), Space O(h)
     */
    public static int maxDepth(TreeNode root) {
        if (root == null) return 0;
        return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Maximum Depth of Binary Tree");
        System.out.println("=".repeat(60));

        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);

        System.out.println("Depth -> " + maxDepth(root));
    }
}

File: basic_dsa_one_hundred_sixty_one.cpp
/**
 * Basic DSA - Maximum Depth of Binary Tree (LeetCode 104)
 * Return the maximum depth (height) of a binary tree.
 */
#include <iostream>
#include <algorithm>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode* l, TreeNode* r) : val(x), left(l), right(r) {}
};

int maxDepth(TreeNode* root) {
    if (!root) return 0;
    return 1 + std::max(maxDepth(root->left), maxDepth(root->right));
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Maximum Depth of Binary Tree\n";
    std::cout << std::string(60, '=') << "\n";

    TreeNode* a = new TreeNode(3,
        new TreeNode(9),
        new TreeNode(20, new TreeNode(15), new TreeNode(7)));
    std::cout << "[3,9,20,null,null,15,7] -> " << maxDepth(a) << "\n";

    TreeNode* b = new TreeNode(1, nullptr, new TreeNode(2));
    std::cout << "[1,null,2] -> " << maxDepth(b) << "\n";

    std::cout << "[] -> " << maxDepth(nullptr) << "\n";
}
